base/src/com/google/idea/blaze/base/sync/workspace/WorkspaceHelper.java

```diff
@@ -155,31 +189,31 @@ private static WorkspacePath getPackagePath(
     return null;
   }
 
+  @VisibleForTesting
+  public static File getExternalSourceRoot(BlazeProjectData projectData) {
+    return new File(projectData.getBlazeInfo().getOutputBase(), "external");
+  }
+
   @Nullable
-  private static synchronized Map<String, WorkspaceRoot> getExternalWorkspaceRoots(
+  private static synchronized WorkspaceRoot getExternalWorkspaceRootsFile(String workspaceName,
       Project project) {
-    if (Blaze.getBuildSystemName(project) == BuildSystemName.Blaze) {
-      return ImmutableMap.of();
+    BlazeProjectData bazelProjectData = getBlazeProjectData(project);
+    if (bazelProjectData == null) {
+      return null;
     }
-    return SyncCache.getInstance(project)
-        .get(WorkspaceHelper.class, WorkspaceHelper::enumerateExternalWorkspaces);
-  }
 
-  @SuppressWarnings("unused")
-  private static Map<String, WorkspaceRoot> enumerateExternalWorkspaces(
-      Project project, BlazeProjectData blazeProjectData) {
-    FileOperationProvider provider = FileOperationProvider.getInstance();
-    File[] children = provider.listFiles(getExternalSourceRoot(blazeProjectData));
-    if (children == null) {
-      return ImmutableMap.of();
+    if (!workspaceRootCache.containsKey(workspaceName)) {
```

Comment:
This is not a primary use case for SyncCache, but I think it's still a good idea to keep it. Otherwise (what happens now) the cache is not cleared on resync. How about this way? Sorry for yet another round but i think it could lead to some bugs

```diff
@@ -201,7 +202,7 @@
     if (bazelProjectData == null) {
       return null;
     }
-
+    var workspaceRootCache = SyncCache.getInstance(project).get(WorkspaceHelper.class, (p, data) -> new ConcurrentHashMap<String, WorkspaceRoot>());
     if (!workspaceRootCache.containsKey(workspaceName)) {
       File externalBase = new File(bazelProjectData.getBlazeInfo().getOutputBase(),
           "external/" + workspaceName);

```

Reply:
Can you elaborate more on the resync case? you mean within one user session when user click on the sync? for that case we don't want to clear the cache right? The external workspace dirs are not going to change. We do want to keep them right? Just trying to understand the case of the cache lifecycle. I thought the cache should stay as long as the `WorkspaceHelper` instance stays. @tpasternak

Reply:
I mean this data's lifetime was previously managed by SyncCache class, which is cleared automatically every time what the sync occurs. I would prefer to keep this behavior. Otherwise we might cause problems when external repositories are renamed etc.

Reply:
hmm if the external workspace root is renamed, the build dep would need to be renamed too right? eg, for some reason the namespace installed is changed in WORKSPACE from:
```
maven_install(
    name = "maven",
    artifacts = [
       //artifacts
    ],
    repositories = [
        "https://repo1.maven.org/maven2",
    ],
)
```
to 
```
maven_install(
    name = "changed_maven",
    artifacts = [
       //artifacts
    ],
    repositories = [
        "https://repo1.maven.org/maven2",
    ],
)
```
Then wherever reference that namespace would need to be changed from
```
@maven//:artifact",
```
to 
```
@changed_maven//:artifact",
```
right? 

Actually i think busting the cache each time when resync is one of the reasons causing the IDE performance slow. But I could be missing some cases here.

Reply:
My understanding is that the primary reason for the improvement was the cache previously being filled with all external repositories at once. Thanks to your change, it now fills lazily, one-by-one. The initial version of your PR also reused CacheSync, which cleared data during each resync. This approach seemed to be working well.

Additionally, it’s not just about renaming but also about cleaning the cache. It shouldn't grow uncontrollably.

Reply:
@tpasternak reverted. Please take a look!

Reply:
Updated. I actually thought abt doing this but it was bit strange to me that we could not utilize the `blazeProjectData` provided by the `SyncCache` and need to keep modifying the cache value(the map). But I understand your concern. Please take another look.